src/protocols/secure_channel/CheckinMessage.cpp

```diff
@@ -0,0 +1,102 @@
+
+    chip::Crypto::HMAC_sha shaHandler;
+    uint8_t nonceWorkBuffer[CHIP_CRYPTO_HASH_LEN_BYTES] = { 0 };
+
+    ReturnErrorOnFailure(shaHandler.HMAC_SHA256(key.As<Aes128KeyByteArray>(), sizeof(Aes128KeyByteArray), appDataStartPtr,
+                                                sizeof(CounterType), nonceWorkBuffer, CHIP_CRYPTO_HASH_LEN_BYTES));
+
+    static_assert(sizeof(nonceWorkBuffer) >= CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES, "We're reading off the end of our buffer.");
+    memcpy(output.data(), nonceWorkBuffer, CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES);
+
+    // In place encryption to save some RAM
+    memcpy(appDataStartPtr + sizeof(CounterType), appData.data(), appData.size());
+
+    uint8_t * micPtr = appDataStartPtr + sizeof(CounterType) + appData.size();
+    ReturnErrorOnFailure(Crypto::AES_CCM_encrypt(appDataStartPtr, sizeof(CounterType) + appData.size(), nullptr, 0, key,
+                                                 output.data(), CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES, appDataStartPtr, micPtr,
+                                                 CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES));
+
+    output.reduce_size(appData.size() + sMinPayloadSize);
+
+    return err;
+}
+
+CHIP_ERROR CheckinMessage::ParseCheckinMessagePayload(Crypto::Aes128KeyHandle & key, ByteSpan & payload, CounterType & counter,
+                                                      MutableByteSpan & appData)
+{
+    VerifyOrReturnError(payload.size() >= sMinPayloadSize, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(payload.size() <= (sMinPayloadSize + sMaxAppDataSize), CHIP_ERROR_INVALID_ARGUMENT);
+
+    CHIP_ERROR err       = CHIP_NO_ERROR;
+    uint16_t appDataSize = GetAppDataSize(payload);
+
+    // To prevent workbuffer usage, appData size needs to be large enough to hold both the appData and the counter
+    VerifyOrReturnError(appData.size() >= sizeof(CounterType) + appDataSize, CHIP_ERROR_INVALID_ARGUMENT);
+
+    ByteSpan nonce         = payload.SubSpan(0, CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES);
+    ByteSpan encryptedData = payload.SubSpan(CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES, sizeof(CounterType) + appDataSize);
+    ByteSpan mic =
+        payload.SubSpan(CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES + sizeof(CounterType) + appDataSize, CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES);
+
+    err = Crypto::AES_CCM_decrypt(encryptedData.data(), encryptedData.size(), nullptr, 0, mic.data(), mic.size(), key, nonce.data(),
+                                  nonce.size(), appData.data());
+
+    ReturnErrorOnFailure(err);
+
+    counter = Encoding::LittleEndian::Get32(appData.data());
+    // Shift to remove the counter from the appData
+    memcpy(appData.data(), sizeof(CounterType) + appData.data(), appDataSize);
+
+    appData.reduce_size(appDataSize);
+    return err;
+}
+
+uint16_t CheckinMessage::GetAppDataSize(ByteSpan & payload)
+{
+    return (payload.size() <= sMinPayloadSize) ? 0 : static_cast<uint16_t>(payload.size() - sMinPayloadSize);
```

Comment:
The why is "Because this will allow us to avoid issues down the line if what we think are invariants change".

In particular, with TCP we're about to have Matter messages that do not have sizes that fit in 16 bits....  For check-in it's probably not going to be an issue, of course.  For now.

Is there any actual efficiency gain to having uint16_t instead of size_t as the return type.  I would expect both to take up the same register in practice.

Reply:
fixed